plugins/qc-firehose/fu-qc-firehose-device.c

```diff
@@ -0,0 +1,1157 @@
+	return fu_qc_firehose_device_read_xml(self, 30000, error);
+}
+
+static gboolean
+fu_qc_firehose_device_write_blocks(FuQcFirehoseDevice *self,
+				   FuChunkArray *chunks,
+				   FuProgress *progress,
+				   GError **error)
+{
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, fu_chunk_array_length(chunks));
+	for (guint i = 0; i < fu_chunk_array_length(chunks); i++) {
+		g_autoptr(FuChunk) chk = NULL;
+
+		/* prepare chunk */
+		chk = fu_chunk_array_index(chunks, i, error);
+		if (chk == NULL)
+			return FALSE;
+		if (!fu_qc_firehose_device_write(self,
+						 fu_chunk_get_data(chk),
+						 fu_chunk_get_data_sz(chk),
+						 error))
+			return FALSE;
+
+		/* update progress */
+		fu_progress_step_done(progress);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gchar *
+fu_qc_firehose_device_convert_to_image_id(const gchar *filename, GError **error)
+{
+	g_autofree gchar *filename_safe = NULL;
+
+	/* sanity check */
+	if (filename == NULL) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_DATA,
+				    "no firmware value");
+		return NULL;
+	}
+	filename_safe = g_strdup(filename);
+	g_strdelimit(filename_safe, "\\", '/');
+	return g_path_get_basename(filename_safe);
+}
+
+static gboolean
+fu_qc_firehose_device_program(FuQcFirehoseDevice *self,
+			      FuFirmware *firmware,
+			      XbNode *xn,
+			      FuProgress *progress,
+			      GError **error)
+{
+	guint64 sector_size = xb_node_get_attr_as_uint(xn, "SECTOR_SIZE_IN_BYTES");
+	guint64 num_sectors = xb_node_get_attr_as_uint(xn, "num_partition_sectors");
+	const gchar *filename = xb_node_get_attr(xn, "filename");
+	g_autofree gchar *filename_basename = NULL;
+	g_autoptr(FuChunkArray) chunks = NULL;
+	g_autoptr(GBytes) blob = NULL;
+	g_autoptr(GBytes) blob_padded = NULL;
+	g_autoptr(XbBuilderNode) bn = xb_builder_node_new("data");
+	g_autoptr(XbBuilderNode) bc = xb_builder_node_insert(bn, xb_node_get_element(xn), NULL);
+	const gchar *names[] = {
+	    "PAGES_PER_BLOCK",
+	    "SECTOR_SIZE_IN_BYTES",
+	    "filename",
+	    "num_partition_sectors",
+	    "physical_partition_number",
+	    "start_sector",
+	    "last_sector",
+	};
+
+	/* sanity check */
+	if ((self->supported_functions & FU_QC_FIREHOSE_FUNCTIONS_PROGRAM) == 0) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "program is not supported");
+		return FALSE;
+	}
+
+	/* skip any empty filenames */
+	filename_basename = fu_qc_firehose_device_convert_to_image_id(filename, error);
+	if (filename_basename == NULL)
+		return FALSE;
+	blob = fu_firmware_get_image_by_id_bytes(firmware, filename_basename, error);
+	if (blob == NULL)
+		return FALSE;
+
+	/* copy across */
+	for (guint i = 0; i < G_N_ELEMENTS(names); i++) {
+		const gchar *value = xb_node_get_attr(xn, names[i]);
+		if (value != NULL)
+			xb_builder_node_set_attr(bc, names[i], value);
+	}
+	if (!fu_qc_firehose_device_write_xml(self, bn, error))
```

Comment:
We write the XML to the device here before we modify it with the correct `num_partition_sectors`.
Technically I think `last_sector` should also be `start_sector + num_partition_sectors - 1`, but I'm not sure if the modem actually cares.

Reply:
Other than that from what I know this all looks very good to me.

Reply:
@Ivanhu5866 it is you I should be thanking -- you're literally the expert here. I'll merge this and rebase my other branch on top -- I'd also appreciate some help with that. **Thanks**!